### DISCUSSION TITLE
Improve MQTT messages of IR Communication

### DISCUSSION THREAD

**Discussion by heikokue** [arendst/Tasmota#17644] [Ideas]
Hello,
currently only the entire IR data telegram is exchanged via the MQTT values `Data` and `DataLSB`. It is therefore difficult for a normal user to establish the relationship between the `Data`/`DataLSB` and the information about which remote control button is behind them.

There are now various databases of all known remote controls, e.g. <https://github.com/probonopd/irdb>.
The device and button numbers behind each remote control button are stored for all devices in these databases.

Here, for example, is an excerpt from an [NEC television](https://github.com/probonopd/irdb/blob/master/codes/NEC/TV/4%2C-1.csv):

|functionname|protocol|device|subdevice|function|
|:----------:|:------:|:----:|:-------:|:------:|
|CHANNEL +   |NEC1    |4     |-1       |0       |
|CHANNEL -   |NEC1    |4     |-1       |1       |
|VOLUME +    |NEC1    |4     |-1       |2       |
|VOLUME -    |NEC1    |4     |-1       |3       |
|POWER       |NEC1    |4     |-1       |8       |

My suggestion would be to extend the MQTT messages of `IRSend` and `IrReceived` in such a way that instead of `Data`/`DataLSB`, the values `device`, `subdevice` and `function` can also be consigned.

This would be much easier for the user to handle the remote buttons via MQTT.

The Coding of the NEC protocol (=`DataLSB`) for example is quite easy (`Address`=`device`, `subdevice` is unused, `Command`=`function`):
![image](https://user-images.githubusercontent.com/74769863/211204958-41b23a5d-0365-421e-9c83-65d1e373fdfd.png)

Due to the large number of IR protocols, this should initially only be implemented for the widespread protocols `NEC`, `RC5` and `RC6` and maybe only in the `tasmota-ir`.

I'm not a software programmer, but I could support you, e.g. as a tester.

Kind Regards
Heiko

**Comment by sfromis** [comment]
With `SetOption58` you can also get "IR Raw data in JSON payload"

**Comment by Jason2866** [comment]
Since flash space is tight in esp8266 every feature added has to be considered.
Adding a conversion tool in the docu would be very helpful, and would not waste precious flash.
Most of the time this info is only needed one time when setting up.
We would be happy to see such a tool, not necessarily in docu site, we could link too.

**Reply by heikokue** [reply]
What do you think about only implement it in the `tasmota-ir` version?

**Comment by heikokue** [comment]
I analyzed the IRremoteESP8266 library and my requested functionality is already implemented as:

- IRsend::encodeNEC(uint16_t address, uint16_t command) and IRrecv::decodeNEC() for the NEC protocol
- IRsend::encodeRC5(const uint8_t address, const uint8_t command,
                           const bool key_released), IRsend::encodeRC6(), IRrecv::decodeRC5() and IRrecv::decodeRC6() for the RC5/RC6 protocols

**Reply by Jason2866** [reply]
You can provide a PR to add the Infos you want enclosed in ifdefs to include only when the switch is set.
So there is no breaking change.
